src/sentry/api/endpoints/artifact_lookup.py

```diff
@@ -0,0 +1,336 @@
+from sentry.api.base import region_silo_endpoint
+from sentry.api.bases.project import ProjectEndpoint, ProjectReleasePermission
+from sentry.api.endpoints.debug_files import has_download_permission
+from sentry.api.serializers import serialize
+from sentry.models import DebugIdArtifactBundle, Distribution, File, Release, ReleaseFile
+from sentry.models.artifactbundle import ArtifactBundleArchive, ReleaseArtifactBundle
+from sentry.models.project import Project
+from sentry.models.releasefile import read_artifact_index
+
+logger = logging.getLogger("sentry.api")
+
+
+# The number of bundles we want to return based on a `debug_id` query.
+MAX_BUNDLES_BY_DEBUG_ID = 4
+
+
+# The number of ArtifactBundles we open up and parse to look for files inside.
+MAX_SCANNED_BUNDLES = 2
+
+
+# The number of files returned by the `get_releasefiles` query
+MAX_RELEASEFILES_QUERY = 10
+
+
+@region_silo_endpoint
+class ProjectArtifactLookupEndpoint(ProjectEndpoint):
+    permission_classes = (ProjectReleasePermission,)
+
+    def download_file(self, file_id, project: Project):
+        rate_limited = ratelimits.is_limited(
+            project=project,
+            key=f"rl:ArtifactLookupEndpoint:download:{file_id}:{project.id}",
+            limit=10,
+        )
+        if rate_limited:
+            logger.info(
+                "notification.rate_limited",
+                extra={"project_id": project.id, "file_id": file_id},
+            )
+            return HttpResponse({"Too many download requests"}, status=429)
+
+        file = File.objects.filter(id=file_id).first()
+
+        if file is None:
+            raise Http404
+
+        try:
+            fp = file.getfile()
+            response = StreamingHttpResponse(
+                iter(lambda: fp.read(4096), b""), content_type="application/octet-stream"
+            )
+            response["Content-Length"] = file.size
+            return response
+        except OSError:
+            raise Http404
+
+    def get(self, request: Request, project: Project) -> Response:
+        """
+        List a Project's Individual Artifacts or Bundles
+        ````````````````````````````````````````
+
+        Retrieve a list of individual artifacts or artifact bundles for a given project.
+
+        :pparam string organization_slug: the slug of the organization to query.
+        :pparam string project_slug: the slug of the project to query.
+        :qparam string debug_id: If set, will query and return all the artifact
+                                  bundles that match one of the given `debug_id`s.
+        :qparam string url: If set, will query and return all the individual
+                             artifacts, or artifact bundles that contain files
+                             that match the `url`. This is using a substring-match.
+        :qparam string release: Used in conjunction with `url`.
+        :qparam string dist: Used in conjunction with `url`.
+
+        :auth: required
```

Comment:
Is this copied from somewhere else?

Reply:
Yes, I started this whole thing as a copy of the DebugFiles endpoint.